File: src/Result.ts
export interface Result<T> {
    unwrap(): T;
}

export class Ok<T> implements Result<T> {
    private value: T;

    constructor(value: T) {
        this.value = value;
    }

    unwrap(): T {
        return this.value;
    }
}

export class Err<E> implements Result<E> {
    private error: E;

    constructor(error: E) {
        this.error = error;
    }

    unwrap(): E {
        throw new Error(String(this.error));
    }
}

File: src/main/java/com/example/Result.java
package com.example;

public abstract class Result<T> {
    
    public abstract T unwrap();
    
    public static class Ok<T> extends Result<T> {
        private T value;

        public Ok(T value) {
            this.value = value;
        }

        @Override
        public T unwrap() {
            return value;
        }
    }

    public static class Err<E> extends Result<E> {
        private E error;

        public Err(E error) {
            this.error = error;
        }

        @Override
        public E unwrap() {
            throw new RuntimeException(error.toString());
        }
    }
}
